### DISCUSSION TITLE
High Ram usage

### DISCUSSION THREAD

**Discussion by bilago** [rustdesk/rustdesk#8651] [General]
### Bug Description

Running the latest nightly, my ram usage for the rustdesk service is 1.47 GB which seems unusual compared to previous builds

### How to Reproduce

Connect to client
wait 30 minutes
observe rustdesk.exe in taskmanager to see the memory in use

### Expected Behavior

Rustdesk should manage itself to keep memory usage lower (assuming it can)

### Operating system(s) on local side and remote side

Windows 11 > Windows 10

### RustDesk Version(s) on local side and remote side

1.2.7 > 1.2.7

### Screenshots

![image](https://github.com/rustdesk/rustdesk/assets/8310293/e0ec45d5-b956-4f3d-9286-c363c6ffc665)

**Comment by 21pages** [comment]
Please try different versions, https://github.com/rustdesk/rustdesk/releases, we wanna confirm which version cause your problem first.

**Comment by rustdesk** [comment]
> compared to previous builds

Are you 100% sure 1.2.6 ok?

**Comment by bilago** [comment]
> > compared to previous builds
> 
> Are you 100% sure 1.2.6 ok?

I went from 1.2.4 to 1.2.7, I didn't update for awhile - sorry for not being clear with my statement. I cannot tell you if 1.2.6 was OK

**Reply by rustdesk** [reply]
Please try out 1.2.6 first, then 1.2.5, it will help us locate the problem quicker. And also, please try out different codecs. and which side has high memory usage, controlling side or controlled side?

**Reply by bilago** [reply]
I will try different versions and report back.

The issue is on the Client side. Host side is only 300mb

**Reply by rustdesk** [reply]
Host side is the controlling side or being controlled side?

**Reply by bilago** [reply]
Controlling side is fine with ram usage, it is the controlled side with high ram usage

**Reply by rustdesk** [reply]
@bilago how about this issue?

**Comment by BomberJet** [comment]
I also experience the problem after the last update
<img width="980" alt="SCR-20240814-pauw" src="https://github.com/user-attachments/assets/2cbb9ef8-3c5e-436c-bc19-7123072556e2">

**Reply by rustdesk** [reply]
1.2.6 is ok for you?

**Reply by rustdesk** [reply]
If 1.2.6 ok for you, could you try out 1.2.6 files in https://github.com/21pages/rustdesk/releases/tag/history ？so that we can get which date exactly we have the problem.

**Reply by BomberJet** [reply]
this is what happens when I stay connected for many hours
<img width="969" alt="SCR-20240816-nlmk" src="https://github.com/user-attachments/assets/9afc7946-341a-4c9f-b322-3c96df867aee">

**Reply by BomberJet** [reply]
I will try version 1.2.6 and let you know in the next few days

**Reply by BomberJet** [reply]
even with version 1.2.6 I encounter the same problem
<img width="994" alt="SCR-20240816-pccp" src="https://github.com/user-attachments/assets/b81a33b0-762e-4f65-bd11-4189fd87a7f9">

**Reply by rustdesk** [reply]
If possible, please try out 1.2.5, 1.2.4 etc.

**Reply by BomberJet** [reply]
The problem is with all versions. However, I have noticed that only on windows 10 is the problem, on windows 11 it is OK.

**Reply by BomberJet** [reply]
windows 10 with 6 hours of connection
<img width="966" alt="win10" src="https://github.com/user-attachments/assets/89bf87bc-24e8-4c08-8f20-dc8604b2cda0">

**Reply by BomberJet** [reply]
windows 11 with 6 hours of connection
<img width="840" alt="win11" src="https://github.com/user-attachments/assets/e032f63a-3e97-46e7-bf32-7cfd89661448">

**Reply by rustdesk** [reply]
Do you mean on all your windows 10 machines (it is unlikey related to win 10 or 11)? What codec are you using on win 10 and 11 when they are being connected?

![image](https://github.com/user-attachments/assets/c9dd2a6f-356b-48fe-a133-86e19b0219ea)

**Reply by BomberJet** [reply]
I only encounter the problem on Windows 10 various versions. The codec is in AUTO (H265) or (H264)

**Reply by rustdesk** [reply]
Could you try out vp9? You may need to restart the `RustDesk` windows service (https://support.cch.com/oss/ml/kb/solution/How-do-I-start-stop-or-restart-a-service) on controlled machine (The Windows 10 machine you mentioned) first so that leak memory can be released.

Also please paste your log (controlled machine) when you see huge memory.
https://github.com/rustdesk/rustdesk/wiki/FAQ#access-logs

**Reply by BomberJet** [reply]
ok, but why on the windows 10 pc? don't i have to select it on the source pc?

**Reply by rustdesk** [reply]
You select vp9 on the source pc.

**Reply by rustdesk** [reply]
Will the memory usage drop after disconnection?

**Reply by BomberJet** [reply]
> Will the memory usage drop after disconnection?

no, I have to restart the service

**Reply by bilago** [reply]
I still am having this issue but I haven't had time to really do any troubleshooting. For now I just set the rustdesk service to auto restart every night at 1am which keeps the ram usage down for my use-case.

**Reply by 21pages** [reply]
@bilago, Could you just try disabling the hardware codec and feedback here? https://github.com/rustdesk/rustdesk/discussions/8651#discussioncomment-10457695

**Reply by BomberJet** [reply]
<img width="983" alt="SCR-20240828-pfhg" src="https://github.com/user-attachments/assets/853a6f2d-0486-44ab-bc38-49f7b0c8b617">
I disabled the hardware codec on the target pc and on the source pc I use the VP9 codec. Now I no longer experience any ram problems. 3 hours of connection 350mb of ram

**Comment by 21pages** [comment]
For users with high ram usage, you can try disabling hardware codecs and restarting the`RustDesk` windows service (https://support.cch.com/oss/ml/kb/solution/How-do-I-start-stop-or-restart-a-service) to see if it still occurs, we want to check whether it's related to hardware codec.

![image](https://github.com/user-attachments/assets/e5b8e5f6-a137-443b-aaea-222f9cd8e34c)

**Reply by BomberJet** [reply]
but does this modification have to be done on the source or target PC?

**Reply by rustdesk** [reply]
This is on the target PC.

**Reply by rustdesk** [reply]
There are two ways to disable hardware codec. The way here on target PC. The other way is selecting `vp9` on source PC.

`restarting the service` are both on target PC.

**Reply by bilago** [reply]
I disabled hardware codec on the target (win 10) machine
Codec "VP9" selected on source (win 11)
Restarted the rustdesk service
Connected from source(win 11) to target (win 10)
Ram usage for rustdesk.exe on target (win 10) is currently at 796,368 K and holding
Will report back if that changes over the course of the day/s

**Reply by bilago** [reply]
Just to note, i don't know if this is expected, but when I change from a single monitor display  to multi monitor display, the ram usage doubles to 1,543,908 K

**Reply by rustdesk** [reply]
Of course, multi monitor display has multiple codecs.

**Reply by 21pages** [reply]
@bilago  Could you seach `new encoder` in `C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\log\server` and find out the exact hardware encoder, now I can't reproduce and I'll  try to reproduce on that encoder.

![1724852724825](https://github.com/user-attachments/assets/2ca0611f-5d56-4f57-a6f3-152d63ae4a98)

**Reply by bilago** [reply]
@21pages 

[2024-08-28 09:37:39.567875 -04:00] INFO [libs\scrap\src\common\codec.rs:131] new encoder: VPX(VpxEncoderConfig { width: 2560, height: 1440, quality: Best, codec: VP9, keyframe_interval: None }), i444: true

**Reply by 21pages** [reply]
Thanks ,but this is the latest log, could you search on the old logs which have the memory problem? h264 and hevc are hardware encoders.

**Reply by rustdesk** [reply]
> Of course, multi monitor display has multiple codecs.

The hardware codec will have less memory usage if it has no memory leak.

**Reply by bilago** [reply]
@21pages [2024-08-14 08:50:09.420496 -04:00] INFO [libs\scrap\src\common\codec.rs:131] new encoder: VRAM(VRamEncoderConfig { device: AdapterDevice { device: 0x22408fe9b78, vendor_id: 32902, luid: 108300 }, width: 2560, height: 1440, quality: Best, feature: FeatureContext { driver: FFMPEG, luid: 108300, api: API_DX11, data_format: H265 }, keyframe_interval: None }), i444: false

**Reply by rustdesk** [reply]
> a single monitor display to multi monitor display

Is your original high ram (H265 codec) related to this?  Will the memory usage drop after disconnection?

**Reply by bilago** [reply]
@rustdesk No, the ram continues to climb overnight with no connection.

**Reply by rustdesk** [reply]
It keeps growing even no connection?

**Reply by bilago** [reply]
Right now with the Software Codec
 1 monitor = 799K . 2 monitor = 1.500K Disconnected: 16K
 
 Switched back to hardware:
 1 monitor = 109K. 2 monitor = 110K Disconnected: 112K and slowly increasing ~1K in the time it took me to write the disconnected usage here. Eventually i will connect again and the starting usage is over 1,000K and remains that way until i stop the service.

**Reply by bilago** [reply]
less than a minute later it's now 115K and I can actively see the usage increasing, and i'm still disconnected.

**Reply by rustdesk** [reply]
Thank you very much. We can confirm it is memory leak related to hardware codec since memory drops after disconnection with software codec, but it does not drop with hardware codec after disconnection.

**Reply by rustdesk** [reply]
How about current memory? 27 minutes elapsed.

**Reply by bilago** [reply]
Well I can't spend any time with it being in a disconnected state, since I'm using the remote session for work. Currently connected and the memory usage is 779K @ single monitor , when it started at 109K  using hardware codec

**Reply by rustdesk** [reply]
Thanks

**Reply by 21pages** [reply]
Which process takes up a lot of memory and keeps leaking memory after disconnecting?

![1724913174315](https://github.com/user-attachments/assets/d16b44e6-97bf-49ea-8a41-fc50ca54c709)

**Reply by bilago** [reply]
The command line entry is empty on windows 10.

Here is something I'm noticing. If I switch between monitor 1 and monitor 2 repeatedly, the ram usage increases by 100,000 K each time. I've switched maybe 8 times in a row and now I'm over 2GB of ram usage

![image](https://github.com/user-attachments/assets/84365f43-5a39-47bc-a279-a71fe66a6608)

**Reply by 21pages** [reply]
> Here is something I'm noticing. If I switch between monitor 1 and monitor 2 repeatedly, the ram usage increases by 100,000 K each time. I've switched maybe 8 times in a row and now I'm over 2GB of ram usage

Thanks, it's the control side or controlled side?

Could you try `Get-WmiObject Win32_Process -Filter "Name = 'rustdesk.exe'" | Select-Object ProcessId, CommandLine` ?
Run it as admin in powershell, you can get the command line with pid.
![image](https://github.com/user-attachments/assets/f7611201-a3b1-4a07-9fbc-c233a455fed6)

**Reply by bilago** [reply]
The ram increasing after monitor switch is happening on the Target (win 10) machine (Controlled side)



ProcessId CommandLine                                       
--------- -----------                                       
    42960 "C:\Program Files\RustDesk\RustDesk.exe" --tray   
     8900 "C:\Program Files\RustDesk\RustDesk.exe"          
    43928 "C:\Program Files\RustDesk\RustDesk.exe" --service
    45260 "C:\Program Files\RustDesk\RustDesk.exe" --server 
    47132 "C:\Program Files\RustDesk\RustDesk.exe" --cm     


  8900 "C:\Program Files\RustDesk\RustDesk.exe"         is the one with high usage currently on the Target machine

**Reply by 21pages** [reply]
8900 is the main window process, it doesn't do encoding, in fact, you can close it when control it.

**Reply by bilago** [reply]
Sorry I was mistaken, i assumed it was 8900 because the commandline was empty like it was in taskmanager.

The actual PID in task manager is 45260 which would be 
   45260 "C:\Program Files\RustDesk\RustDesk.exe" --server

**Reply by 21pages** [reply]
Thanks, this helps.

**Reply by 21pages** [reply]
> The ram increasing after monitor switch is happening on the Target (win 10) machine (Controlled side)

Reproduced this way, I'll try to fix it.

**Comment by 21pages** [comment]
It happens on Intel gpu as controlled side and can be fixed by update ffmpeg, it'll be fixed soon.

**Comment by rustdesk** [comment]
Try out this please, we fixed this controlled side ffmpeg encoder leak, https://github.com/rustdesk/rustdesk/releases/nightly, also fixed a gpu leak on controlling side. So please update both sides.

**Comment by AddisZone** [comment]
Run over night without connections. The service seems to be the cause of mem leak:

![mem-usage](https://github.com/user-attachments/assets/e9cafd79-7de0-4c7f-ad50-268226bf698f)

Y axis is MByte. 
W: PS Get-Process.WorkingSet64 
P: PS Get-Process.PrivateMemorySize64

**Reply by rustdesk** [reply]
So is it ok now, or not?

**Reply by AddisZone** [reply]
"rustdesk --service" mem usage is rising. So IMHO not ok because I see no reason why it requires more and more mem. It's not that much raise in memory usage like before but it should not happen.

**Reply by AddisZone** [reply]
The powershell script continues to run and diligently collects data every 10 seconds. :)

**Reply by rustdesk** [reply]
Do you have RDP session now?

**Reply by AddisZone** [reply]
Yes. The collected data will also include this RDP session. There is an additional "rustdesk --cm" process now.

**Reply by rustdesk** [reply]
Could you close this RDP and see if the memory continue to grow? I see there is a leak related to RDP.

**Reply by AddisZone** [reply]
Unfortunately that's not possible today as I'm working remotely from home office on that machine. It will end in about 7 hours. Then it will continue measurement without connection over weekend.

**Reply by rustdesk** [reply]
Please try out this https://github.com/rustdesk/rustdesk/releases/tag/nightly, I just fixed a leak.

BTW, it has nothing to do with RDP.

**Comment by AddisZone** [comment]
The result of mem usage til this morning. The 'rustdesk --service' process is continuously raising with about 1.3MB/h

![mem-usage-2](https://github.com/user-attachments/assets/8e1623f6-9cc5-40cc-9e76-80c36a395968)

**Reply by rustdesk** [reply]
The fix https://github.com/rustdesk/rustdesk/commit/179b562472a1edf9d68e97a2ed11926c8779b881 was merged on `2024-09-13 07:42 UTC` time, so your build does not have this fix. Please use the latest nightly.

**Comment by AddisZone** [comment]
- memory usage recording started short after Rustdesk update
- at about 15:15 remote session to other controlled pc for few minutes

primary Y axis (MByte):
W:     $process.WorkingSet64 / 1MB
P:     $process.PrivateMemorySize64 / 1MB
T:     $process.Threads.Count
PkW:   $process.PeakWorkingSet64 / 1MB
noPgS: $process.NonpagedSystemMemorySize64 / 1MB
Pg:    $process.PagedMemorySize64 / 1MB
PkPg:  $process.PeakPagedMemorySize64 / 1MB
PgS:   $process.PagedSystemMemorySize64 / 1MB

secondary Y axis (count):
H:     $process.HandleCount

![mem-usage-4](https://github.com/user-attachments/assets/7106e6dd-cb17-4d2f-a145-f1e5e962ca10)

rustdesk --service: 
looks ok

rustdesk --server:
looks ok

rustdesk --tray:
looks ok

rustdesk:
- long decay in W memory usage reduction after connection closed
- P memory usage stays doubled after connection closed
- about 70 Handles more after connection closed
- Threads before connection 28..31 but after connection closed 33..35

I'm not familiar with the meaning of [Windows PS memory values](https://superuser.com/questions/618686/private-bytes-vs-working-set-in-process-explorer) .  So it can be the case that W and P aren't critical.

**Reply by AddisZone** [reply]
Memory usage behavior when controlled by other PC (=> service) couldn't be checked at that moment.

**Comment by GowthamKudupudi** [comment]
lol, rustdesk is leaking memory after all the rust's bandwagon! On RPi3B 
<img alt="Initial RAM" src="https://lh3.googleusercontent.com/pw/AP1GczPJa7gPXjg_yLON2GWJ_WWWdnCprykPp_YayJPMJ2RA3qElxknBH5zOiPxipWSZ_Heayv-dj5NX8flgk08jnuFCG4ZuQWbS41nsonV05RDJ5VDtDWHH_rYTzxXFO8vOqWh4gZiNjqMFvjpKVnBSr2H3=w1262-h1594-s-no-gm?authuser=0" width="300"/><img alt="30min RAM" src="https://lh3.googleusercontent.com/pw/AP1GczNFE_tGZrET4ofcYqsyRjGk4Gxhn_C6Xml0oxZVs7bHB9kSN2Mo7cqQM3ISoc4pHekLeJotIymVyRDqJApHlHs9_QwFOFVnajsPGCH6KaaGpg3UO9CwhmXtxdjUjPBNrlCA7TwVKdGn3PvwUk7OD_4e=w1342-h1594-s-no-gm?authuser=0" width="300"/>
after 45min RPi3B hungs up
